bokeh/models/tests/test_plots.py

```diff
@@ -159,3 +161,73 @@ def test_plot_raises_error_if_toolbar_and_tools_are_set():
 def test_plot_with_no_title_specified_creates_an_empty_title():
     plot = Plot()
     assert plot.title.text == ""
+
+
+def test_plot__scale_classmethod():
+    assert isinstance(Plot._scale("auto"), LinearScale)
+    assert isinstance(Plot._scale("linear"), LinearScale)
+    assert isinstance(Plot._scale("log"), LogScale)
+    assert isinstance(Plot._scale("categorical"), CategoricalScale)
+    with pytest.raises(ValueError):
+        Plot._scale("malformed_type")
+
+
+def test_plot_x_mapper_type_kwarg_sets_x_scale():
+    plot = Plot(x_mapper_type="linear")
+    assert isinstance(plot.x_scale, LinearScale)
+
+
+def test_plot_y_mapper_type_kwarg_sets_y_scale():
+    plot = Plot(y_mapper_type="log")
+    assert isinstance(plot.y_scale, LogScale)
+
+
+def test_plot_raises_error_if_x_mapper_type_and_x_scale_are_set():
+    with pytest.raises(ValueError):
+        Plot(x_mapper_type="linear", x_scale=LinearScale())
+
+
+def test_plot_raises_error_if_y_mapper_type_and_y_scale_are_set():
+    with pytest.raises(ValueError):
+        Plot(y_mapper_type="log", y_scale=LogScale())
+
+
+def test__check_required_scale_has_scales():
+    plot = Plot()
+    check = plot._check_required_scale()
+    assert check == []
+
+
+def test__check_required_scale_missing_scales():
+    plot = Plot(x_scale=None, y_scale=None)
+    check = plot._check_required_scale()
+    assert check != []
+
+
+def test__check_compatible_scale_and_ranges_compat_numeric():
+    plot = Plot(x_scale=LinearScale(), x_range=Range1d())
+    check = plot._check_compatible_scale_and_ranges()
+    assert check == []
+
+
+    plot = Plot(y_scale=LogScale(), y_range=DataRange1d())
+    check = plot._check_compatible_scale_and_ranges()
+    assert check == []
+
+
+def test__check_compatible_scale_and_ranges_compat_factor():
+    plot = Plot(x_scale=CategoricalScale(), x_range=FactorRange())
+    check = plot._check_compatible_scale_and_ranges()
+    assert check == []
+
+
+def test__check_compatible_scale_and_ranges_incompat_numeric_scale_and_factor_range():
+    plot = Plot(x_scale=LinearScale(), x_range=FactorRange())
+    check = plot._check_compatible_scale_and_ranges()
+    assert check != []
+
+
+# def test__check_compatible_scale_and_ranges_incompat_factor_scale_and_numeric_range():
```

Comment:
There's some weirdness here because `CategoricalScale` is a subclass of `LinearScale`. Examples will work because a categorical_scale will use the linear_scale compute/invert methods if the passed values are numbers.

We need to decide we should prevent this situation on the python side though.